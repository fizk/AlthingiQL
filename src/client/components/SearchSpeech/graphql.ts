import gql from 'graphql-tag';

export const searchIssueSpeeches = gql`
    query searchIssueSpeeches ($assembly: Int! $issue: Int! $query: String!) {
        SearchIssueSpeeches(issue: $issue, assembly: $assembly, query: $query) {
            id
            assembly {id}
            issue {id}
            congressmanType
            period {from to}
            congressman {
                id
                name
                avatar {
                    templateSrc
                }
                party {
                    id
                    name
                    color
                }
            }
            text
        }
    }
`;
