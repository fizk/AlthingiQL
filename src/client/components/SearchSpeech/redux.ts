
const SPEECH_SEARCH = {
    BEGIN: 'SPEECH_SEARCH:BEGIN',
    END: 'SPEECH_SEARCH:END',
    ERROR: 'SPEECH_SEARCH:ERROR',
    CLEAR: 'SPEECH_SEARCH:CLEAR',
};

const speechSearchActionBegin = () => ({
    type: SPEECH_SEARCH.BEGIN,
});

const speechSearchActionEnd = result => ({
    type: SPEECH_SEARCH.END,
    result,
});

const speechSearchActionError = error => ({
    type: SPEECH_SEARCH.ERROR,
    error,
});

export const speechSearchClearAction = () => ({
    type: SPEECH_SEARCH.CLEAR,
});

const searchQuery = `
    query searchIssueSpeeches ($assembly: Int! $issue: Int! $query: String!) {
        SearchIssueSpeeches(issue: $issue, assembly: $assembly, query: $query) {
            id
            assembly {id}
            issue {id}
            congressmanType
            period {from to}
            congressman {
                id
                name
                party {
                    id
                    name
                    color
                }
            }
            text
        }
    }
`;



export const speechSearchAction = (assembly, issue, query) => {
    return dispatch => {
        dispatch(speechSearchActionBegin());

        const headers = new Headers();
        headers.append('Content-Type', 'application/json');

        const req = new Request(__GRAPHQL_SERVER__, {
            method: 'POST',
            body: JSON.stringify({
                query: searchQuery,
                variables: {assembly, issue, query},
            }),
            headers: headers,
        });

        return fetch(req)
            .then(response => response.json())
            .then(json => {
                if(json.hasOwnProperty('errors')) {
                    dispatch(speechSearchActionError(json.errors))
                } else {
                    dispatch(speechSearchActionEnd(json.data.SearchIssueSpeeches));
                }

            })
            .catch(error => dispatch(speechSearchActionError(error)))
    }
};

export default (state = {result: [], loading: false, error: false}, action) => {
    switch (action.type) {
        case SPEECH_SEARCH.BEGIN:
            return {
                ...state,
                loading: true,
                error: false
            };
            break;
        case SPEECH_SEARCH.END:
            return {
                result: action.result,
                loading: false,
                error: false
            };
            break;
        case SPEECH_SEARCH.ERROR:
            return {
                ...state,
                loading: false,
                error: action.error
            };
            break;
        case SPEECH_SEARCH.CLEAR:
            return {
                result: [],
                loading: false,
                error: false
            };
            break;
        default:
            return state;
            break;
    }
}
