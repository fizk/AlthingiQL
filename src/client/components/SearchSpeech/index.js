import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom'
import {OptionsWithKeyBinding} from '../../elements/Form';
import {speechSearchAction, speechSearchClearAction} from './redux';
import {SpeechSearchResult} from '../../elements/SearchResult';

export default class SearchSpeech extends React.Component {
    static propTypes = {
        assembly: PropTypes.number,
        issue: PropTypes.number,
        result: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.string,
            assembly: PropTypes.shape({
                id: PropTypes.number,
            }),
            issue: PropTypes.shape({
                id: PropTypes.number,
            }),
            text: PropTypes.string,
            period: PropTypes.shape({
                from: PropTypes.string,
                to: PropTypes.string,
            }),
            iteration: PropTypes.string,
            type: PropTypes.string,
            congressmanType: PropTypes.string,
            congressman: PropTypes.shape({
                id: PropTypes.number,
                name: PropTypes.string,
                avatar: PropTypes.shape({
                    templateSrc: PropTypes.string,
                }),
                party: PropTypes.shape({
                    id: PropTypes.number,
                    name: PropTypes.string,
                    color: PropTypes.string,
                })
            }),
        })),
        isSearching: PropTypes.bool,
        onSearch: PropTypes.func,
        onClear: PropTypes.func,
    };

    static defaultProps = {
        assembly: undefined,
        issue: undefined,
        result: [],
        isSearching: false,
        onSearch: () => {},
        onClear: () => {},
    };

    constructor(props) {
        super(props);

        this.handleOnChange = this.handleOnChange.bind(this);
        this.handleOnSelect = this.handleOnSelect.bind(this);
        this.handleOnClear = this.handleOnClear.bind(this);

        this.state = {
            redirect: undefined
        }
    }

    handleOnChange(value) {
        if (value.trim() !== '') {
            this.props.onSearch(value);
        }
    }

    handleOnSelect(value) {
        this.props.onClear();
        this.setState({redirect: value})
    }

    handleOnClear() {
        this.props.onClear()
    }

    render() {
        return (
            <Fragment>
                {this.state.redirect && <Redirect push={true} to={`/loggjafarthing/${this.state.redirect.assembly.id}/thingmal/${this.state.redirect.issue.id}/raedur/${this.state.redirect.id}`} />}
                <OptionsWithKeyBinding
                    isSearching={this.props.isSearching}
                    onClear={this.handleOnClear}
                    onChange={this.handleOnChange}
                    onSelect={this.handleOnSelect}>
                    {this.props.result.map(result => (
                        <SpeechSearchResult key={`speech-${result.id}`} value={result} />
                    ))}
                </OptionsWithKeyBinding>
            </Fragment>
        )
    }
}

const mapStateToProps = state => {
    return {
        result: state.speechSearch.result,
        isSearching: state.speechSearch.loading,
    }
};

const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        onSearch: query => dispatch(speechSearchAction(ownProps.assembly, ownProps.issue,  query)),
        onClear: () => dispatch(speechSearchClearAction())
    }
};

const SearchSpeechWithStore = connect(mapStateToProps, mapDispatchToProps)(SearchSpeech);

export {SearchSpeechWithStore}
