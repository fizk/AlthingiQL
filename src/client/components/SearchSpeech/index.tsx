import * as React from "react";
import { Fragment } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import { OptionsWithKeyBinding } from "../../elements/Form";
import { speechSearchAction, speechSearchClearAction } from "./redux";
import { SpeechSearchResult } from "../../elements/SearchResult";

type SearchSpeechProps = {
    assembly?: number,
    issue?: number,
    result?: {
        id?: string,
        assembly?: {
            id?: number
        },
        issue?: {
            id?: number
        },
        text?: string,
        period?: {
            from?: string,
            to?: string
        },
        iteration?: string,
        type?: string,
        congressmanType?: string,
        congressman?: {
            id?: number,
            name?: string,
            avatar?: {
                templateSrc?: string
            },
            party?: {
                id?: number,
                name?: string,
                color?: string
            }
        }
    }[],
    isSearching?: boolean,
    onSearch?: (...args: any[]) => any,
    onClear?: (...args: any[]) => any
};

type SearchSpeechState = {
    redirect: any | undefined
};

export default class SearchSpeech extends React.Component<SearchSpeechProps, SearchSpeechState> {
    static defaultProps = {
        assembly: undefined,
        issue: undefined,
        result: [],
        isSearching: false,
        onSearch: () => {},
        onClear: () => {}
    };
    constructor(props) {
        super(props);
        this.handleOnChange = this.handleOnChange.bind(this);
        this.handleOnSelect = this.handleOnSelect.bind(this);
        this.handleOnClear = this.handleOnClear.bind(this);
        this.state = {
            redirect: undefined
        };
    }
    handleOnChange(value) {
        if (value.trim() !== "") {
            this.props.onSearch(value);
        }
    }
    handleOnSelect(value) {
        this.props.onClear();
        this.setState({ redirect: value });
    }
    handleOnClear() {
        this.props.onClear();
    }
    render() {
        return (
            <Fragment>
                {this.state.redirect && (
                    <Redirect
                        push={true}
                        to={`/loggjafarthing/${
                            this.state.redirect.assembly.id
                        }/thingmal/${this.state.redirect.issue.id}/raedur/${
                            this.state.redirect.id
                        }`}
                    />
                )}
                <OptionsWithKeyBinding
                    isSearching={this.props.isSearching}
                    onClear={this.handleOnClear}
                    onChange={this.handleOnChange}
                    onSelect={this.handleOnSelect}
                >
                    {this.props.result.map(result => (
                        <SpeechSearchResult
                            key={`speech-${result.id}`}
                            value={result}
                        />
                    ))}
                </OptionsWithKeyBinding>
            </Fragment>
        );
    }
}
const mapStateToProps = state => {
    return {
        result: state.speechSearch.result,
        isSearching: state.speechSearch.loading
    };
};
const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        onSearch: query =>
            dispatch(
                speechSearchAction(ownProps.assembly, ownProps.issue, query)
            ),
        onClear: () => dispatch(speechSearchClearAction())
    };
};
const SearchSpeechWithStore = connect(
    mapStateToProps,
    mapDispatchToProps
)(SearchSpeech);
export { SearchSpeechWithStore };
