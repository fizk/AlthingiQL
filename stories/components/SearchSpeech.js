import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import {SearchSpeech} from '../../src/client/components/SearchSpeech';
import '../../public/stylesheets/storybook.css';


const results = [
    {
        "id": "20161207T170115",
        "assembly": {
            "id": 146
        },
        "issue": {
            "id": 1
        },
        "congressmanType": null,
        "period": {
            "from": "2016-12-07T06:01:15.000Z",
            "to": "2016-12-07T06:03:01.000Z"
        },
        "congressman": {
            "id": 1158,
            "name": "Björt Ólafsdóttir",
            "party": {
                "id": 44,
                "name": "Björt framtíð",
                "color": "92278f"
            }
        },
        "text": "hugmyndafræði sem liggur að baki.Ég **hef** hins vegar rætt **það** margoft **í** þessum stól, og [...]  skattastyrknum **í** þessu samhengi þá skil ég einfaldlega ekki forsendurnar, ég verð alveg að viðurkenna **það**, eða þá [...] ,  menntakerfið og hvað **það** heitir, sem svo stóriðjan **í** sínum rekstri notar eins og allir aðrir hér **í** [...] . þingmanni fyrir að beina athygli minni að þessum svokallaða skattastyrk. Ég verð að viðurkenna að ég **hef** [...]  ekki lesið þessar tilteknu blaðsíður. Þær eru nokkrar reyndar, þær blaðsíðurnar sem ég **hef** ekki lesið"
    },
    {
        "id": "20161207T141917",
        "assembly": {
            "id": 146
        },
        "issue": {
            "id": 1
        },
        "congressmanType": "fjármála- og efnahagsráðherra",
        "period": {
            "from": "2016-12-07T03:19:17.000Z",
            "to": "2016-12-07T03:20:33.000Z"
        },
        "congressman": {
            "id": 652,
            "name": "Bjarni Benediktsson",
            "party": {
                "id": 35,
                "name": "Sjálfstæðisflokkur",
                "color": "0000ff"
            }
        },
        "text": "stöndum ágætlega við þau **í** þessu fjárlagafrumvarpi. Eins og ég **hef** oft sagt má auðvitað halda því fram að [...]  greiðsluþátttökuna, ég **hef** nefnt einstaka fjárfestingarverkefni; **það** má meðal annars horfa til [...] -Evrópuríkin sem eru að reyna að skrapa saman **í** jöfnuð **í** ríkisfjármálum.  Flest þeirra eru með mikinn [...]  halla.Seðlabankinn hefur kallað eftir því að menn sýni aðhald. **Það** er akkúrat einn [...]  meginþráðurinn **í** málflutningi mínum. Varðandi loforð um að halda áfram að byggja upp  tel ég að við"
    },
    {
        "id": "20161222T155039",
        "assembly": {
            "id": 146
        },
        "issue": {
            "id": 1
        },
        "congressmanType": null,
        "period": {
            "from": "2016-12-22T04:50:39.000Z",
            "to": "2016-12-22T04:52:48.000Z"
        },
        "congressman": {
            "id": 690,
            "name": "Katrín Jakobsdóttir",
            "party": {
                "id": 23,
                "name": "Vinstri hreyfingin - grænt framboð",
                "color": "ff0000"
            }
        },
        "text": "fyrir  þessar spurningar. Hún spurði  um sameiningar háskóla. Sjálf **hef** ég verið þeirrar skoðunar að [...]  samstarf  háskóla **í** öllu falli sé mjög mikilvægt og ég kom til að mynda á verkefni á sínum tíma **í** [...]  ráðuneyti menntamála sem var samstarfsnet hinna opinberu háskóla. Hugsunin á bak við **það** var að gera [...]  allt samstarf auðveldara og þar með að leggja **í** raun og veru upp með að auðveldara væri fyrir þessa [...]  skóla að kenna t.d.  saman, halda saman utan um kennslu **í** námsgreinum **í** ljósi þess að við erum með sjö"
    },
];

storiesOf('SearchSpeech', module)
    .add('search result', () => (
        <SearchSpeech assembly={1}
                      issue={2}
                      result={results}
                      onSearch={action('onSearch')}
                      onClear={action('onClear')} />
    ))
    .add('searching', () => (
        <SearchSpeech assembly={1}
                      issue={2}
                      result={[]}
                      isSearching={true}
                      onSearch={action('onSearch')}
                      onClear={action('onClear')} />
    ));
