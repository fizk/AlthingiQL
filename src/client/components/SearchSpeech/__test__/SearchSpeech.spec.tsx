import * as React from 'react';
import SearchSpeech from '../index';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';

describe('<SearchSpeech />', () => {
    test('default props', () => {
        const tree = renderer.create(
            <Router >
                <SearchSpeech />
            </Router>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});


