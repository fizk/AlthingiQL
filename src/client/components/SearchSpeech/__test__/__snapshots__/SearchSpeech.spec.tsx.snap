// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchSpeech /> data 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <input
    className="input input--md input--search"
    onChange={[Function]}
    onFocus={[Function]}
    type="search"
  />
  <div
    className="options-list__container"
    style={
      Object {
        "position": "absolute",
        "zIndex": 1,
      }
    }
  >
    <div
      className="options-list"
    >
      <div
        className="options-list__item options-list__item--active"
        onClick={[Function]}
      >
        <div
          className="speech-card"
        >
          <article
            className="congressman"
          >
            <header
              className="congressman__avatar"
            >
              <div
                className="avatar"
                role="image"
                style={
                  Object {
                    "backgroundImage": "url(/path/to/image.jpg)",
                  }
                }
                title="LjÃ³smynd af name"
              />
            </header>
            <div
              className="congressman__party"
            >
              <div
                className="badge"
                style={
                  Object {
                    "backgroundColor": "#123123",
                  }
                }
                title="name"
              />
            </div>
            <div
              className="congressman__body"
            >
              <h4
                className="headline-3 headline-3--ellipsis"
              >
                name
              </h4>
              <h5
                className="headline-4"
              >
                congressman type
              </h5>
            </div>
          </article>
          <div
            className="speech-card__meta"
          >
            <span
              className="speech-card__type"
            >
              type
            </span>
            <span
              className="speech-card__iteration"
            >
              1
            </span>
            <time
              className="speech-card__date"
            >
              2001 M01 01 11:00:00
            </time>
            <time
              className="speech-card__date"
            >
              11:00:00
            </time>
          </div>
          <div
            className="speech-card__text"
          >
            <p>
              text
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchSpeech /> default props 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <input
    className="input input--md input--search"
    onChange={[Function]}
    onFocus={[Function]}
    type="search"
  />
</div>
`;
